
# Chapter 5: Characterising the Time-Resolved Myeloid Cell Atlas
### *(not yet refactored)*

I am in the process of refactoring this repository to better reflect my abilities in the present day. As of yet, this chapter has not been refactored. Please see the timestamps to find my latest work.


### Chapter Introduction
scRNAseq data requires substantial low-level analysis as described in Chapters 3 & 4. However, the ultimate purpose of generating this type of data is to make new biological discoveries at single cell resolution. Approaches to facilitate biological interpretation and discovery through scRNAseq are flexible and numerous. In this chapter, I first verify that the novel dataset contains the expected cell types and states, the markers of which have become increasingly available in the literature over the course of this PhD. After establishing this, I take the typical first step in exploratory analysis: grouping transcriptionally similar cells through unsupervised clustering. Then, I investigate the cluster dynamics after SCI through trajectory analysis. From there, I characterise the clusters of interest, drawing heavily from previous findings and resources such as gene ontology (GO) and PanglaoDB, and finally present a summary of proposed myeloid cell dynamics after SCI.
